xdsl/interpreters/experimental/pdl.py

```diff
@@ -0,0 +1,296 @@
+            interpreter.set_values(matcher.assignment.items())
+            self.rewriter = rewriter
+
+            for rewrite_impl_op in pdl_rewrite_op.body.ops:
+                interpreter.run(rewrite_impl_op)
+
+            interpreter.pop_scope()
+
+        rewriter = AnonymousRewritePattern(rewrite)
+
+        PatternRewriteWalker(
+            rewriter, apply_recursively=False).rewrite_module(input_module)
+
+        return ()
+
+    @impl(pdl.OperationOp)
+    def run_operation(self, interpreter: Interpreter, op: pdl.OperationOp,
+                      args: tuple[Any, ...]) -> tuple[Any, ...]:
+        assert op.opName is not None
+        op_name = op.opName.data
+        op_type = self.ctx.get_optional_op(op_name)
+
+        if op_type is None:
+            raise InterpretationError(
+                f'Could not find op type for name {op_name} in context')
+
+        attribute_value_names = [
+            avn.data for avn in op.attributeValueNames.data
+        ]
+
+        # How to deal with operand_segment_sizes?
+        # operand_values, attribute_values, type_values = args
+
+        operand_values = interpreter.get_values(op.operandValues)
+        for operand in operand_values:
+            assert isinstance(operand, SSAValue)
+
+        attribute_values = interpreter.get_values(op.attributeValues)
+
+        for attribute in attribute_values:
+            assert isinstance(attribute, Attribute)
+
+        type_values = interpreter.get_values(op.typeValues)
+
+        for type_value in type_values:
+            assert isinstance(type_value, MLIRType)
+
+        attributes = dict(zip(attribute_value_names, attribute_values))
+
+        result_op = op_type.create(operands=operand_values,
+                                   result_types=type_values,
+                                   attributes=attributes)
+
+        return result_op,
+
+    @impl(pdl.ReplaceOp)
+    def run_replace(self, interpreter: Interpreter, op: pdl.ReplaceOp,
+                    args: tuple[Any, ...]) -> tuple[Any, ...]:
+        rewriter = self.rewriter
+
+        old, = interpreter.get_values((op.opValue, ))
+
+        if op.replOperation is not None:
+            new_op, = interpreter.get_values((op.replOperation, ))
+            rewriter.replace_op(old, new_op)
+        elif len(op.replValues):
+            new_vals = interpreter.get_values(op.replValues)
+            rewriter.replace_op(old, new_ops=[], new_results=list(new_vals))
+        else:
+            assert False, 'Unexpected ReplaceOp'
+
+        return ()
+
+    @impl(ModuleOp)
+    def run_module(self, interpreter: Interpreter, op: ModuleOp,
+                   args: tuple[Any, ...]) -> tuple[Any, ...]:
+        ops = op.ops
+        if len(ops) != 1 or not isinstance(ops[0], pdl.PatternOp):
+            raise InterpretationError(
+                'Expected single pattern op in pdl module')
+        return self.run_pattern(interpreter, ops[0], args)
```

Comment:
It feels weird that we define how a `builtin.module` is interpreted here, as it is not specific to PDL. I think a module just has a "interpret line-by-line" semantics and depending on what op is hit, the interpreter dispatches.
For this experimental PR I am fine with this.

Reply:
Yeah I've been thinking about this a lot, and I don't know what to do. The module is special in that it is the executable, and depending on what it's composed of has very different execution semantics. In Toy, it should run the "main" function. In PDL it should go for the pattern. I'm sure other dialects will be similar. I'm not sure how to specify "I want this interpreter to use this main runtime implementation". Feels out of scope of this PR.

Reply:
I agree, was just a nit here.
I think it makes sense to revisit this when we look at crafting larger more general interpreters